[sw/silicon_creator] Add OTBN ROM driver

jon-flatley opened the pull request:
Adds an OTBN driver implementation with unittests. This driver is heavily based off of sw/device/lib/dif/dif_otbn*, introduced in commit 1c7cf8f8d24e ("[otbn] Initial DIF") with slimmed down error checking and irq, init, and restart routines pruned.

jon-flatley commented on sw/device/silicon_creator/lib/drivers/otbn.h:
```diff
@@ -0,0 +1,155 @@
+#endif
+
+typedef struct otbn {
+  mmio_region_t base_addr;
```
Ah, didn't catch abs_mmio going in, I'll update accordingly.

cfrantz replied:
Yes, abs_mmio is the now the way to perform MMIO from within mask ROM code.

cfrantz commented on sw/device/silicon_creator/lib/drivers/otbn.c:
```diff
@@ -0,0 +1,113 @@
+}
+
+size_t otbn_get_dmem_size_bytes(const otbn_t *otbn) {
+  return OTBN_DMEM_SIZE_BYTES;
```
Yep.  In general, the Mask ROM really shouldn't have functions which just return a constant value - we should just use the constant value.

bbhanushali commented:
@cfrantz Can you PTAL at this review?

alphan commented on sw/device/silicon_creator/lib/drivers/otbn.h:
```diff
@@ -0,0 +1,155 @@
+ * can be seen at the same time. This enum gives the individual bits that may be
+ * set for different errors.
+ */
+typedef enum otbn_err_bits {
+  kOtbnErrBitsNoError = 0,
+  /** Load or store to invalid address. */
+  kOtbnErrBitsBadDataAddr = (1 << 0),
+  /** Instruction fetch from invalid address. */
+  kOtbnErrBitsBadInsnAddr = (1 << 1),
+  /** Call stack underflow or overflow. */
+  kOtbnErrBitsCallStack = (1 << 2),
+  /** Illegal instruction execution attempted */
+  kOtbnErrBitsIllegalInsn = (1 << 3),
+  /** LOOP[I] related error */
+  kOtbnErrBitsLoop = (1 << 4),
+  /** Error seen in Imem read */
+  kOtbnErrBitsFatalImem = (1 << 5),
+  /** Error seen in Dmem read */
+  kOtbnErrBitsFatalDmem = (1 << 6),
+  /** Error seen in RF read */
+  kOtbnErrBitsFatalReg = (1 << 7)
+} otbn_err_bits_t;
```
I think we should add static assertions for these to make sure that they match `OTBN_ERR_BITS_*_BIT`

alphan reviewed (commented):
Thanks @jon-flatley! I left some comments, probably for things already covered by other reviews. I think the most important point is abs_mmio.

jon-flatley commented on sw/device/silicon_creator/lib/drivers/otbn.c:
```diff
@@ -0,0 +1,113 @@
+                             size_t mem_size) {
+  // The overflow check below assumes/requires two unsigned inputs.
+  return (len_bytes % sizeof(uint32_t) == 0 &&
+          offset_bytes % sizeof(uint32_t) == 0 &&
```
I've added a TODO to switch this once #6112 is resolved.

cfrantz replied:
In general, I support migrating this to alignment helper functions.

Realistically, in this case, the modulo operation with a constant power-of-two divisor (ie: sizeof(uint32_t), aka 4) is going to get optimized by the compiler into a bitwise-AND with the value 3.

My concern in sec_mmio.c was because the divisor was not constant, but even then, my concern might have been misplaced.  `__udivdi3` is for long ints (aka uint64_t on rv32i, whereas in sec_mmio.c, the modulus is a 32/32 divide).

I'm not recommending a change here, just elaborating on my original reasoning in sec_mmio, and why my concern may have been misplaced.

jon-flatley commented:
Thanks for the detailed reviews! I believe I've addressed all of your comments and added some things of note

- Changed all functions taking void * types to now take uint32_t * to naturally solve some alignment concerns.
- Remaining alignment checks waiting on #6112 before removing mod operator
- Added assertions for error values
- Removed functions that returned constant values and replaced them with constants.
- Restructured unittests to match guidelines

cfrantz commented on sw/device/silicon_creator/lib/drivers/otbn.h:
```diff
@@ -0,0 +1,134 @@
+ * @param len number of words to copy.
+ * @return `kErrorOtbnInvalidArgument` if len is invalid, `kErrorOk` otherwise.
+ */
+rom_error_t otbn_imem_read(uint32_t offset_bytes, uint32_t *dest, size_t len);
```
Is there a use case for this function in the ROM?

jon-flatley replied:
I can't imagine a use case for this anywhere. I can purge it; it's easy enough to reintroduce if needed.

cfrantz commented on sw/device/silicon_creator/lib/drivers/otbn.c:
```diff
@@ -0,0 +1,118 @@
+  return kErrorOk;
+}
+
+rom_error_t otbn_start(unsigned int start_addr) {
```
Should start_addr be `uint32_t`?

jon-flatley replied:
Done.

alphan commented on sw/device/silicon_creator/lib/drivers/otbn_unittest.cc:
```diff
@@ -0,0 +1,194 @@
+#include "sw/device/silicon_creator/lib/drivers/otbn.h"
+
+#include <array>
+#include <gmock/gmock-matchers.h>
```
nit: can be removed

alphan reviewed (approved):
Thanks @jon-flatley, just some nits.
